core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -272,7 +272,10 @@ private DocumentMapper merge(DocumentMapper mapper, boolean updateAllTypes) {
             if (Version.indexCreated(indexSettings).onOrAfter(Version.V_2_0_0_beta1) && mapper.type().equals(mapper.parentFieldMapper().type())) {
                 throw new IllegalArgumentException("The [_parent.type] option can't point to the same type");
             }
-            if (mapper.type().contains(".") && !PercolatorService.TYPE_NAME.equals(mapper.type())) {
+            if (Version.indexCreated(indexSettings).onOrAfter(Version.V_2_0_0_beta1) && typeContainsIllegalDot(mapper)) {
+                throw new IllegalArgumentException("mapping type name [" + mapper.type() + "] must not contain '.' in it");
+            }
+            if (typeContainsIllegalDot(mapper)) {
```

Comment:
maybe put this in an `if else` clause? For me this makes it clearer what is pre 2.0 and what is 2.0 behaviour.